### DISCUSSION TITLE
How to push the server to 100% CPU?

### DISCUSSION THREAD

**Discussion by WuBingzheng** [tokio-rs/tokio#5991] [Q&A] [answered]
I wrote a simple echo service based on tokio. It spawns a task for each connect received.

Then I wrote a client program that can specify N tasks to request the echo service simultaneously. Each task establishes a persistent connection and sequentially makes several requests. 

I wanted to use this client to maximize the CPU usage of the echo service, but I was never able to do so.

I'm using an M1 MacBook. The echo service automatically started 8 threads. The client program runs at the same computer.

Here is the cpu usage for different number of client tasks:

```
tasks of  | CPU usage(%) of
client    | echo server
----------+----------------
  1       |  34
  2       |  54
  3       |  72
  4       |  99
  5       |  116
  6       |  127
  7       |  134
  8       |  137
  9       |  137
 10       |  137
```

You can see than the CPU stops at 137% with 8 client-tasks.

However I wish to push the server and the client to use 400% CPU both, to run out of the total 8 cores.

Is it possible? How to make it?

----

I set the threads of echo server, and test it:

```
threads of  | CPU usage(%) of
echo server | echo server
------------+----------------
 1          | 86
 2          | 130
 3          | 132
 4          | 136
 5+         | 136
```

You can see the echo server can not use 100% CPU even for 1 thread.
2 threads is better than 1 thread. But after 2+ threads, more threads do not help to increase the echo-server's performance.

----

I found a similar discussion: https://github.com/tokio-rs/tokio/discussions/5779, but there is no further progress. 

Here is the echo server's code:

```
use tokio::net::TcpListener;
use tokio::io::{AsyncReadExt, AsyncWriteExt};

#[tokio::main]
async fn main() {
    let listener = TcpListener::bind("127.0.0.1:2345").await.unwrap();
    loop {
        let (mut socket, _) = listener.accept().await.unwrap();
        tokio::spawn(async move {
            let mut buf = Vec::with_capacity(200);
            while let Ok(len) = socket.read_buf(&mut buf).await {
                if len == 0 {
                    break;
                }
                if let Err(_) = socket.write_all(&buf).await { // echo the request
                    break;
                }
                buf.clear();
            }
        });
    }
}
```

and the client's code:

```
use std::env;
use std::time::Instant;
use tokio::net::TcpStream;
use tokio::io::{AsyncWriteExt, AsyncReadExt, self};

#[tokio::main]
async fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        println!("Usage: {} server-address #task #request", args[0]);
        return;
    }
    let ntask = args[2].parse::<usize>().unwrap();
    let nreq = args[3].parse::<usize>().unwrap();

    let mut handlers = Vec::new();
    for _ in 0..ntask {
        let mut conn = TcpStream::connect(&args[1]).await.unwrap();
        handlers.push(tokio::spawn(async move {
            let start = Instant::now();
            for _ in 0..nreq {
                request(&mut conn).await.unwrap();
            }
            println!("task time: {:?}", start.elapsed());
        }));
    }

    for h in handlers {
        let _ = h.await;
    }
}

async fn request(stream: &mut TcpStream) -> io::Result<()> {
    stream.write_all("hello, tokio!".as_bytes()).await?;

    let mut buf = [0; 100];
    stream.read(&mut buf).await?;
    Ok(())
}
```

**Comment by wathenjiang** [comment]
You may set the number of worker threads to a greater number, to make more worker threads are running instead of IO waiting.

And if your machine are running other process (the desktop computer always have), the tokio process cant take all cpu time because of OS preemptive schedule mechanism.

**Reply by WuBingzheng** [reply]
I change to run the test on a Linux server, with 4 cores. There is less other process such as desktop.
I set the number of threads to 8, for the echo-server and bench-client both. So there are 16 threads in total.
But it gives the same result.

**Reply by wathenjiang** [reply]
It is because the above code will always waiting for IO, even if we add threads, though CPU can not be busy.

**Reply by WuBingzheng** [reply]
According to what you said, if I keep increasing the client request frequency by adding more concurrent connections from the client side, then the server's IO waiting time should be reduced, then the server CPU usage should gradually increase until the entire machine's CPU is saturated. However, based on test results, when the number of client tasks reaches 8, the server CPU stops growing.

**Reply by wathenjiang** [reply]
You can just call a cpu busy function when every TCP stream have incoming bytes, the CPU could be very busy.

Actually, if you use HTTP protocol instead of TCP protocol in this case, the CPU would be very busy.

**Reply by WuBingzheng** [reply]
My original purpose was not to saturate the machine's CPU, but to test the performance limit of my service program. My service program is relatively simple, more complex than the echo server but probably simpler than HTTP protocol.

Now when I do bench testing, the QPS of the service doesn't go up, but the CPU is not saturated. So I want to know, is this caused by tokio's own mechanism, or am I using tokio incorrectly, or is my bench-client method wrong?

**Reply by Noah-Kennedy** [reply]
@WuBingzheng I'm taking a look at this.

What value do you use for `nreq`?

**Reply by Noah-Kennedy** [reply]
Nevermind, I'm just using 10 million to ensure things run for a bit.

**Reply by Noah-Kennedy** [reply]
I'm not as familiar with MacOS specifically, but in Linux pumping IO through the system like this will result in SoftIRQ eating a ton of CPU. Something similar is probably happening for you.

What does your systemwide CPU look like when you run these tests?

As a side note, I'm almost done writing up a thing on this sort of issue. I'll post it in here and tailor it slightly when I'm done.

**Reply by Noah-Kennedy** [reply]
I'm interested in both userspace and kernelspace CPU usage.

**Reply by WuBingzheng** [reply]
I am not familiar with Mac either, I do not know how to see SoftIRQ in Mac. So I test this at Linux with a 4-core CPU. Both the echo-server and bench-client have 4 threads.

Here is the `top` result:

```
top - 02:47:02 up 17:23,  4 users,  load average: 0.37, 0.24, 0.12
Tasks: 138 total,   1 running, 137 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3.7 us, 17.0 sy,  0.0 ni, 70.8 id,  0.0 wa,  0.0 hi,  8.5 si,  0.0 st
MiB Mem :   1963.4 total,    391.0 free,    225.5 used,   1346.8 buff/cache
MiB Swap:   2048.0 total,   2038.6 free,      9.4 used.   1556.2 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  13227 wub       20   0  275024   1324   1192 S  75.3   0.1   0:05.37 tokio-echo
  13234 wub       20   0  410216   1336   1204 S  75.0   0.1   0:05.42 simple-bench
```

Here is the `mpstat` result:

```
Average:     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
Average:     all    4.45    0.00   16.99    0.00    0.00   10.09    0.00    0.00    0.00   68.46
Average:       0    5.32    0.00   15.97    0.00    0.00   10.50    0.00    0.00    0.00   68.20
Average:       1    3.70    0.00   15.83    0.00    0.00    9.47    0.00    0.00    0.00   71.01
Average:       2    4.42    0.00   16.67    0.00    0.00   10.77    0.00    0.00    0.00   68.14
Average:       3    4.35    0.00   19.32    0.00    0.00    9.66    0.00    0.00    0.00   66.67
```

You can see that the %soft does not eat too much CPU, and there is 60+% total idle.

**Reply by Noah-Kennedy** [reply]
Could you use `cargo flamegraph` to get a few flamegraphs?

I'm seeing SoftIRQ eat my entire CPU, which is exactly what's to be expected here:
![vanilla-16t-1024c-flamegraph](https://github.com/tokio-rs/tokio/assets/9875622/5e47ee37-2894-4a1b-b8cb-eec9d42c7715)

**Reply by Noah-Kennedy** [reply]
Trying to figure out what is happening different in your Linux environment...

**Reply by Noah-Kennedy** [reply]
Also, for context on what I'm seeing from htop:
![htop](https://github.com/tokio-rs/tokio/assets/9875622/ca355200-d026-4088-ad03-5c22270cbc4e)

**Reply by Noah-Kennedy** [reply]
Hmm, let me try toning down my thread count here.

**Reply by Noah-Kennedy** [reply]
It's possible that I'm getting contended rx queues and you aren't due to fewer threads.

**Reply by Noah-Kennedy** [reply]
That made no difference to my flamegraph, which actually makes sense.

**Reply by Noah-Kennedy** [reply]
curious what you've got.

**Reply by Noah-Kennedy** [reply]
Also, what throughput are you getting through loopback?

**Reply by WuBingzheng** [reply]
loopback throughput:

```
03:29:02 AM     IFACE   rxpck/s   txpck/s    rxkB/s    txkB/s   rxcmp/s   txcmp/s  rxmcst/s   %ifutil
03:29:03 AM        lo 222756.00 222756.00  14139.79  14139.79      0.00      0.00      0.00      0.00
03:29:03 AM    enp0s3      0.00      0.00      0.00      0.00      0.00      0.00      0.00      0.00
03:29:03 AM    enp0s8      1.00      5.00      0.06      2.56      0.00      0.00      0.00      0.00
```

**Reply by WuBingzheng** [reply]
<img width="2559" alt="image" src="https://github.com/tokio-rs/tokio/assets/12245980/38805c1b-84fe-4145-8547-15d8cf812613">

**Reply by Noah-Kennedy** [reply]
Can you try invoking `echo 1 > /proc/sys/kernel/perf_event_paranoid` and running again?

**Reply by Noah-Kennedy** [reply]
We need to resolve those `[unknown]` stacks.

**Reply by Noah-Kennedy** [reply]
Might also need `echo 0 > /proc/sys/kernel/kptr_restrict`

**Reply by WuBingzheng** [reply]
Maybe because I did not run it as `root`.
I have to fix some ENV to run `cargo` as root.
I will post the result some hours later.
Thanks very much.

**Reply by Noah-Kennedy** [reply]
No problem!

**Reply by Noah-Kennedy** [reply]
I might be in bed when you respond. If so, I'll respond in the morning.

**Reply by WuBingzheng** [reply]
![flamegraph](https://github.com/tokio-rs/tokio/assets/12245980/ff8c2c90-4005-4dd9-ace9-492e6a63a03f)

**Reply by WuBingzheng** [reply]
Did you run the bench-client at the same machine with echo-server?
How many cores does your CPU have?
How many threads did you set for both echo-server and bench-client?
@Noah-Kennedy

**Reply by Noah-Kennedy** [reply]
@WuBingzheng I ran both on the same machine. My CPU has 16 physical, 32 logical. I've tried various configurations, including giving both 16 cores (which was when I captured my flamegraph).

Loopback TCP has entirely different performance characteristics from TCP going over a NIC due to how it works with systems like SoftIRQ, so I kept both on the same machine so as to better recreate your setup.

**Reply by Noah-Kennedy** [reply]
Also, it does appear that you are spending your time in softirq.

Could you post a screenshot of htop?

**Reply by Noah-Kennedy** [reply]
while you run

**Reply by WuBingzheng** [reply]
<img width="1574" alt="image" src="https://github.com/tokio-rs/tokio/assets/12245980/7b5c5f29-8483-40a1-a072-c2a28acfd710">

**Reply by WuBingzheng** [reply]
when I set 8 threads for both server and client:

<img width="1640" alt="image" src="https://github.com/tokio-rs/tokio/assets/12245980/6b1e1a20-d6c8-4850-900d-c72ada4b11a7">

**Reply by WuBingzheng** [reply]
There should be 8+1=9 threads, but I do not know why there are only 7 threads in the `htop` 's result above.

Here is the `ps` result:

```
wub@ubuntu:~$ ps -L -p 20516
    PID     LWP TTY          TIME CMD
  20516   20516 pts/0    00:00:00 tokio-ch
  20516   20518 pts/0    00:00:06 tokio-runtime-w
  20516   20519 pts/0    00:00:05 tokio-runtime-w
  20516   20520 pts/0    00:00:06 tokio-runtime-w
  20516   20521 pts/0    00:00:06 tokio-runtime-w
  20516   20522 pts/0    00:00:04 tokio-runtime-w
  20516   20523 pts/0    00:00:06 tokio-runtime-w
  20516   20524 pts/0    00:00:05 tokio-runtime-w
  20516   20525 pts/0    00:00:00 tokio-runtime-w
```

**Reply by Noah-Kennedy** [reply]
What are the specs of the Linux server? Specifically, I'm interested in CPU, kernel version, and if its a VM or not.

**Reply by WuBingzheng** [reply]
```
wub@ubuntu:~$ lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  4
  On-line CPU(s) list:   0-3
Vendor ID:               GenuineIntel
  Model name:            13th Gen Intel(R) Core(TM) i5-13500
    CPU family:          6
    Model:               191
    Thread(s) per core:  1
    Core(s) per socket:  4
    Socket(s):           1
    Stepping:            2
    BogoMIPS:            4991.99
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc
                         cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase bmi1 avx2 bmi2 invpcid r
                         dseed clflushopt md_clear flush_l1d arch_capabilities
```

```
wub@ubuntu:~$ uname -a
Linux ubuntu 5.15.0-83-generic #92-Ubuntu SMP Mon Aug 14 09:30:42 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
wub@ubuntu:~$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
```

Yes, it's a VM, running in VirtualBox in Windows, which runs at intel i5 CPU.

**Reply by Noah-Kennedy** [reply]
Oh, are you still using 8 connections?

**Reply by Noah-Kennedy** [reply]
Try 1024

**Reply by WuBingzheng** [reply]
<img width="1543" alt="image" src="https://github.com/tokio-rs/tokio/assets/12245980/04fae880-b9c4-4dcd-a428-c1079b908f54">

```
top - 01:56:05 up 1 day, 14:06,  6 users,  load average: 2.69, 0.86, 0.30
Tasks: 144 total,   1 running, 143 sleeping,   0 stopped,   0 zombie
%Cpu(s): 10.4 us, 40.2 sy,  0.0 ni, 29.4 id,  0.0 wa,  0.0 hi, 19.9 si,  0.0 st
MiB Mem :   1963.4 total,    338.4 free,    249.0 used,   1376.0 buff/cache
MiB Swap:   2048.0 total,   2036.3 free,     11.7 used.   1528.5 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  21088 wub       20   0  546088   3044   2480 S 143.9   0.2   1:37.36 tokio-ch
  21098 wub       20   0  614060   3860   2756 S 142.9   0.2   1:37.19 simple-bench
     14 root      20   0       0      0      0 I   0.3   0.0   2:27.04 rcu_sched
  20763 root      20   0       0      0      0 I   0.3   0.0   0:08.53 kworker/3:0-events
```

**Reply by WuBingzheng** [reply]
```

Average:     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
Average:     all    9.75    0.00   39.67    0.00    0.00   20.16    0.00    0.00    0.00   30.42
Average:       0   11.11    0.00   39.72    0.00    0.00   21.39    0.00    0.00    0.00   27.78
Average:       1    9.62    0.00   39.55    0.00    0.00   21.02    0.00    0.00    0.00   29.81
Average:       2    8.84    0.00   40.19    0.00    0.00   19.01    0.00    0.00    0.00   31.96
Average:       3    9.42    0.00   39.21    0.00    0.00   19.19    0.00    0.00    0.00   32.18
```

**Reply by WuBingzheng** [reply]
I try 10000 tasks, then the CPU reaches to 100% almost!

<img width="1607" alt="image" src="https://github.com/tokio-rs/tokio/assets/12245980/2d9ba527-bac9-4b0e-b320-07e3ed1b96cf">

**Reply by Noah-Kennedy** [reply, marked as answer]
I forgot you were only running 8 connections.

You'll never be able to get enough throughput from 8 connections without resorting to the dark arts of network programming to actually use much CPU

**Reply by Noah-Kennedy** [reply]
That CPU is gonna be like 70% softirq

**Reply by Noah-Kennedy** [reply]
probably more

**Reply by Noah-Kennedy** [reply]
If you want more throughput for **something which is strictly IO-bound like this**, its better to use SO_REUSEPORT and a pool of current_thread runtime threads which are pinned to different cores, with sockets pinned to core rx queues using SO_INCOMING_CPU

**Reply by Noah-Kennedy** [reply]
if you want more than that then you can look into ebpf or dpdk for the stupidly high-throughput echo server

**Reply by WuBingzheng** [reply]
In my very first test at Mac, the CPU utils did not change after the number of tasks growing bigger than 8. You can find the result in the original post.
So I thought the number of tasks did not matter, so I did not try much bigger number of tasks.

- for Mac, 8 tasks and 1000 tasks are almost same;
- for Linux, 1000 tasks is much better than 8 tasks.

**Reply by WuBingzheng** [reply]
Thanks for your advices, I will try them later.

Thanks very much!

**Reply by Noah-Kennedy** [reply]
> In my very first test at Mac, the CPU utils did not change after the number of tasks growing bigger than 8. You can find the result in the original post.
> 
> So I thought the number of tasks did not matter, so I did not try much bigger number of tasks.
> 
> 
> 
> - for Mac, 8 tasks and 1000 tasks are almost same;
> 
> - for Linux, 1000 tasks is much better than 8 tasks.

that's very interesting, i wonder what the macos network stack looks like and where it's bottlenecked.

given my past experiences with macos internals though i'm not really surprised.

**Reply by WuBingzheng** [reply]
> ... and a pool of current_thread runtime threads which are pinned to different cores ...

Is this equivalent to [monoio](https://github.com/bytedance/monoio) and [Glommio](https://github.com/DataDog/glommio)?

**Reply by Noah-Kennedy** [reply]
i marked one of my statements as the answer, but we can keep this thread going

**Reply by Noah-Kennedy** [reply]
Both of those use strategies similar to this, however both are built around io_uring rather than epoll. io_uring seems to decently enough for TCP-based network programming use cases so long as you fall into one of the use cases where a runtime-per-thread model is acceptable. Unfortunately, this actually limits its usefulness quite a bit.

Speaking of io_uring, I need to do some testing again on io_uring's network performance sometime to see if it has improved since I last was messing with it, and also do a bunch of work on tokio-uring.
